cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkRuntimeUtils.java

```diff
@@ -232,10 +258,68 @@ private void handleCompleted(boolean failure) {
         if (failure && driverService instanceof SparkDriverService) {
           ((SparkDriverService) driverService).stopWithoutComplete();
         } else {
-          driverService.stopAndWait();
+          driverService.stop();
         }
+
+        // Wait for the spark execution context to stop to make sure everything related
+        // to the spark execution is shutdown properly before returning.
+        // When this method returns, the driver's main thread returns, which will have system services shutdown
+        // and process terminated. If system services stopped concurrently with the execution context, we may have some
+        // logs and tasks missing.
+        Uninterruptibles.awaitUninterruptibly(secStopLatch, 30, TimeUnit.SECONDS);
+      }
+    };
+  }
+
+  /**
+   * Adds a {@link EventLoggingListener} to the Spark context.
+   *
+   * @param runtimeContext the {@link SparkRuntimeContext} for connecting to CDAP services
+   * @return A {@link Closeable} which should be called when the Spark application completed
+   */
+  public static Closeable addEventLoggingListener(SparkRuntimeContext runtimeContext) throws IOException {
+    // If upload event logs is not enabled, just return a dummy closeable
+    if (!runtimeContext.getCConfiguration().getBoolean(Constants.AppFabric.SPARK_EVENT_LOGS_ENABLED)) {
+      return () -> { };
+    }
+
+    SparkConf sparkConf = new SparkConf();
+    sparkConf.set("spark.eventLog.enabled", Boolean.toString(true));
+    sparkConf.set("spark.eventLog.compress", Boolean.toString(true));
+
+    ProgramRunId programRunId = runtimeContext.getProgramRunId();
+
+    File eventLogDir = Files.createTempDirectory("spark-events").toFile();
+    LOG.debug("Spark events log directory for {} is {}", programRunId, eventLogDir);
+
+    // EvevntLogginginListener is Scala package private[spark] class.
+    // However, since JVM bytecode doesn't really have this type of cross package private support, the class
+    // is actually public class. This is why we can access it in Java code.
+    EventLoggingListener listener = new EventLoggingListener(Long.toString(System.currentTimeMillis()), Option.empty(),
+                                                             eventLogDir.toURI(), sparkConf) {
+      @Override
+      public void onApplicationStart(SparkListenerApplicationStart event) {
+        // Rewrite the application start event with identifiable names based on the program run id such that
+        // it can rely back to the program run.
```

Comment:
nit: "refer back" instead of "rely back"

Reply:
fixed